go/store/types/inlineblob.go

```diff
@@ -0,0 +1,127 @@
+// Copyright 2019 Liquidata, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package types
+
+import (
+	"context"
+	"encoding/base64"
+	"encoding/binary"
+	"fmt"
+	"math"
+
+	"github.com/liquidata-inc/dolt/go/store/hash"
+)
+
+type InlineBlob []byte
+
+func (v InlineBlob) Value(ctx context.Context) (Value, error) {
+	return v, nil
+}
+
+func (v InlineBlob) Equals(other Value) bool {
+	v2, ok := other.(InlineBlob)
+	if !ok {
+		return false
+	}
+
+	vlen := len(v)
+	if vlen != len(v2) {
+		return false
+	}
+
+	for i := 0; i < vlen; i++ {
+		if v[i] != v2[i] {
+			return false
+		}
+	}
+
+	return true
+}
+
+func (v InlineBlob) Less(nbf *NomsBinFormat, other LesserValuable) (bool, error) {
+	if v2, ok := other.(InlineBlob); ok {
+		vlen := len(v)
+		v2len := len(v2)
+		if vlen == v2len {
+			for i := 0; i < vlen; i++ {
+				b1 := v[i]
+				b2 := v2[i]
+
+				if b1 != b2 {
+					return b1 < b2, nil
+				}
+			}
+			return false, nil
+		}
+		return vlen < v2len, nil
+	}
+	return InlineBlobKind < other.Kind(), nil
+}
+
+func (v InlineBlob) Hash(nbf *NomsBinFormat) (hash.Hash, error) {
+	return getHash(v, nbf)
+}
+
+func (v InlineBlob) WalkValues(ctx context.Context, cb ValueCallback) error {
+	return nil
+}
+
+func (v InlineBlob) WalkRefs(nbf *NomsBinFormat, cb RefCallback) error {
+	return nil
+}
+
+func (v InlineBlob) typeOf() (*Type, error) {
+	return InlineBlobType, nil
+}
+
+func (v InlineBlob) Kind() NomsKind {
+	return InlineBlobKind
+}
+
+func (v InlineBlob) valueReadWriter() ValueReadWriter {
+	return nil
+}
+
+func (v InlineBlob) writeTo(w nomsWriter, nbf *NomsBinFormat) error {
+	byteLen := len(v)
+	if byteLen > math.MaxUint16 {
+		return fmt.Errorf("InlineBlob has length %v when max is %v", byteLen, math.MaxUint16)
+	}
+	byteLenSl := make([]byte, 2)
+	binary.BigEndian.PutUint16(byteLenSl, uint16(byteLen))
+
+	err := InlineBlobKind.writeTo(w, nbf)
+	if err != nil {
+		return err
+	}
+
+	w.writeBytes(byteLenSl)
+	w.writeBytes(v)
+	return nil
+}
+
+func (v InlineBlob) valueBytes(nbf *NomsBinFormat) ([]byte, error) {
+	byteLen := len(v)
```

Comment:
I'm not completely sure. Looking at `UUID`, we're not adding the `UUIDKind` in `valueBytes`, but we do in `writeTo`. Looking at `Int` now, I see that `valueBytes` does make use of `writeTo`, so it is writing the `NomsKind`. It seems like one way may be a bug versus the other way, but I'm not 100% sure.

Reply:
Seems like valueBytes is definitely expecting a kind prefix. See `bool.go` which comes from noms, etc.

Reply:
Found a bug in uuid.go though it sounds like!

Reply:
Ahh `bool` is also appending the kind. How big of a bug do you think this may be in `UUID`? Part of the work I was going to do on the `types` package was to eliminate the two functions in favor of just one, but I've no idea how _fixing_ the bug will affect things. I'm also surprised as to how this hasn't been caught yet.

Reply:
I think it's probably not too big a bug and I think fixing it is decently unlikely to cause any problems. Persistence, for example, doesn't use this path AFAIK.

Reply:
Ohh okay, good to know! I'll just bundle that change here since it's a small fix.